### PR TITLE
W02 d01 iterators with tests

### PR DESCRIPTION
**Pull request by tgaff**
Adding final exercises with TESTS
* would love some comments here
* I intend to instruct them not to look at the test code

**this PR merges into w02-d01-iterators**

### CODE REVIEW

File: week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+### running tests
+
+Run the tests individually as you work on each challenge.  For example to run
+the tests for myMap:
+
+```bash
+mocha test/test-myMap.js
+```
+
+This will test against your solution in `myMap.js`.
```
Comment (nathanallen) [inline week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/README.md]: Consider adding some example output here, and what it means.

File: week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/test/test-myEach.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+var myEach = require('../myEach');
+
+describe('myEach', function() {
+  // sample data
+  beforeEach(function() {
+    testArr = [0, 1, 100, 1000];
+  });
+
+  it("takes and calls a callback function", function testCallback() {
```
Comment (nathanallen) [inline week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/test/test-myEach.js]: it takes a function as an argument, and calls/executes that function

File: week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/test/test-myEach.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    var resultingArray = [];
+    myEach(testArr, function(item) {
+      resultingArray.push(item);
+    });
+    // compare elements in the result to expected array
+    console.log('       results: ', resultingArray);
+    expect(resultingArray).to.have.members([0, 1, 100, 1000]);
+  });
+
+  it("passes each item in the array to the callback, even non-integers", function testArrayPassing() {
```
Comment (nathanallen) [inline week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/test/test-myEach.js]: each value

File: week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/test/test-myEach.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    myEach(complexTestArr, function(item) {
+      resultingArray.push(item);
+    });
+
+    console.log('       results: ', resultingArray);
+    expect(resultingArray).to.have.members(complexTestArr);
+  });
+
+
+  it("passes each index in the array to the callback", function testEachIndex() {
```
Comment (nathanallen) [inline week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/test/test-myEach.js]: as the second argument

File: week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/test/test-myEach.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    myEach(testArr, function(_item, index) {
+      resultingArray.push(index);
+    });
+    // compare elements in the result to expected array
+    console.log('       results: ', resultingArray);
+    expect(resultingArray).to.have.members([0, 1, 2, 3]);
+  });
+
+
+  it("passes the entire array to the callback", function testArrayPassing() {
```
Comment (nathanallen) [inline week-02-working-with-objects/day-01-iterators/dusk-iterators/exercises_c/test/test-myEach.js]: as the third argument

**Comment by nathanallen** [comment]
This looks good!

1) the exercises folder needs to be moved into a separate lab repo.

2) Make sure to spend a moment reading the MDN documentation together. Don't assume students know how to read it yet.
  -- e.g. "optional argument" notation (it's not an array!)
  -- what arguments do the callbacks have: i.e. currentValue, index, array (forEach)

3) Run the tests together, and look at the output together.
 -- What does it mean to run a test? What's being run? (Students will put method calls inside their js because they assume they need to call the functions explicitly).
 -- What does a failing test look like / what does a passing test look like?

**Comment by tgaff** [comment]
Thanks @nathanallen

**Comment by tgaff** [comment]
You're absolutely right that this should be in another repo.  It *grew* here and I didn't think to transplant it.

**Pull request merged**